clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,1957 @@
+                            end_column: self.column as u32,
+                        },
+                        e: LexerError::UnterminatedUTF8Encoding,
+                    });
+                    return error_string.to_string();
+                }
+                '"' => {
+                    self.add_diagnostic(PlacedError {
+                        span: Span {
+                            start_line: self.line as u32,
+                            start_column: self.column as u32,
+                            end_line: self.line as u32,
+                            end_column: self.column as u32,
+                        },
+                        e: LexerError::UnterminatedUTF8Encoding,
+                    });
+                    self.add_diagnostic(PlacedError {
+                        span: Span {
+                            start_line,
+                            start_column,
+                            end_line: start_line,
+                            end_column: start_column,
+                        },
+                        e: LexerError::NoteToMatchThis('}'),
+                    });
+                    return error_string.to_string();
+                }
+                _ => {
+                    if self.next.is_ascii_hexdigit() {
+                        code.push(self.next);
+                    } else {
+                        self.add_diagnostic(PlacedError {
+                            span: Span {
+                                start_line: self.line as u32,
+                                start_column: self.column as u32,
+                                end_line: self.line as u32,
+                                end_column: self.column as u32,
+                            },
+                            e: LexerError::IllegalCharUTF8Encoding(self.next),
+                        });
+                    }
+                }
+            }
+        }
+    }
+
+    fn read_utf8_string(&mut self) -> Vec<Vec<u8>> {
+        let start_line = self.last_line as u32;
+        let start_column = self.last_column as u32;
+        let mut data: Vec<Vec<u8>> = vec![];
+        let mut escaped = false;
+        let mut advance = true;
+        self.read_char();
+        loop {
+            if escaped {
+                match self.next {
+                    '\\' => data.push("\\".to_string().into_bytes()),
+                    '\"' => data.push("\"".to_string().into_bytes()),
+                    'n' => data.push("\n".to_string().into_bytes()),
+                    't' => data.push("\t".to_string().into_bytes()),
+                    'r' => data.push("\r".to_string().into_bytes()),
+                    '0' => data.push("\0".to_string().into_bytes()),
+                    'u' => {
+                        let encode_str = self.read_utf8_encoding();
+                        let unicode_char = {
+                            let u = u32::from_str_radix(&encode_str, 16).unwrap();
```

Comment:
What if `encode_str` is more than 8 characters long, and thus can't fit into `u32`?

Reply:
Fixed in cda9ec15c05757dc68f4ec1a2614cc53e8177817